dispatcher.go

```diff
@@ -188,12 +197,18 @@ func (d *Dispatcher) ClientConfig(service string) transport.ClientConfig {
 // Procedures returns a list of services and procedures that have been
 // registered with this Dispatcher.
 func (d *Dispatcher) Procedures() []transport.Procedure {
+	if d.routerMiddleware != nil {
+		return d.routerMiddleware.Procedures(d.table)
+	}
 	return d.table.Procedures()
 }
 
 // Choose picks a handler for the given request or returns an error if a
 // handler for this request does not exist.
 func (d *Dispatcher) Choose(ctx context.Context, req *transport.Request) (transport.HandlerSpec, error) {
+	if d.routerMiddleware != nil {
+		return d.routerMiddleware.Choose(ctx, req, d.table)
```

Comment:
Use the same pattern here that we use for outbounds and handlers.

Create an `ApplyRouterMiddleware` function which wraps a `Router` with a
`RouterMiddleware` and use that.

You may have to make a private version of it that works with RouteTables.